src/SafeAreaContext.tsx

```diff
@@ -59,40 +68,61 @@ export function SafeAreaProvider({
         nativeEvent: { frame: nextFrame, insets: nextInsets },
       } = event;
 
-      if (
-        // Backwards compat with old native code that won't send frame.
-        nextFrame &&
-        (nextFrame.height !== frame.height ||
-          nextFrame.width !== frame.width ||
-          nextFrame.x !== frame.x ||
-          nextFrame.y !== frame.y)
-      ) {
-        setFrame(nextFrame);
-      }
-
-      if (
-        !insets ||
-        nextInsets.bottom !== insets.bottom ||
-        nextInsets.left !== insets.left ||
-        nextInsets.right !== insets.right ||
-        nextInsets.top !== insets.top
-      ) {
-        setInsets(nextInsets);
-      }
+      setFrame((frame) => {
+        if (
+          // Backwards compat with old native code that won't send frame.
+          nextFrame &&
+          (nextFrame.height !== frame.height ||
+            nextFrame.width !== frame.width ||
+            nextFrame.x !== frame.x ||
+            nextFrame.y !== frame.y)
+        ) {
+          return nextFrame;
+        } else {
+          return frame;
+        }
+      });
+
+      setInsets((insets) => {
+        if (
+          !insets ||
+          nextInsets.bottom !== insets.bottom ||
+          nextInsets.left !== insets.left ||
+          nextInsets.right !== insets.right ||
+          nextInsets.top !== insets.top
+        ) {
+          return nextInsets;
+        } else {
+          return insets;
+        }
+      });
     },
-    [frame, insets],
+    [setFrame, setInsets],
   );
 
+  const latestInsets = useLatest(insets);
+  const latestFrame = useLatest(frame);
+  const getLatestWindowMetrics = React.useCallback(() => {
```

Comment:
Personally I had these use cases:
* A reanimated component that doesn't re-render. Using `useSafeAreaInsets` subscribes to re-render when values change.
* In reanimated, I use the imperative `measure` util, so needed an imperative way to get safe insets at that time.

Reply:
I wouldn't worry too much about re-rendering. I think if you need an imperative method - this can be implemented user-side via reading from the context, setting it on some refs, and then adding your own `addListener` context

Reply:
https://legacy.reactjs.org/docs/hooks-faq.html#how-to-read-an-often-changing-value-from-usecallback

Reply:
Thanks very much!

Got it will get that dep array empty. 

Thanks!